### PR TITLE
WL#21201 Snap-together Magnetic blocks

### PR DESCRIPTION
**Pull request by Menithal**
https://worklist.net/21201

## Simple Snapping Magnetic Blocks

Added two scripts
- makeBlocks.js - Creates a 3x3 array of cloneable, randomly colored pastel cubes you can snap together after placing into world
- magneticBlock.js - Blocks with snap the nearest entity within 0.5 - 2 the scale of the object that is derezed.

#### Basic

1. Run makeBlocks script
2. 3x3 square of entities should rez, script should go away
3. Grabbing an magnetic block should clone it
4. You can place it anywhere else. Calling it entity B
5. Grab another block, and place it near a face of B.

Expected: should hear a snap sound and the block aligning with the other

#### Advanced

1. Run makeBlocks script
2. 3x3 square of entities should rez, script should go away
3. Grabbing an magnetic block should clone it
4. Resize the cloned block to aproximately 3x the size
5. Clone another block
6. Attempt to snap cloned block to rescaled one. 

Expected: Should not snap

#### Not in Scope of test:

- MagneticBlocks will simply snap by matching rotations, it will not calculate any special angles to find nearest faces. 
- Floating point issues may cause there sometimes to be visible seams in the entity.
- Having many small cubes in a 3d matrix may cause any far in enough cubes pushed snap to interior faces
- Having Large entities clip through smaller ones and vice versa in various configurations.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: fe19b5511ce5b1f04c42e6db7369c1c9f075c289 available at: https://gist.github.com/e1058c73d102567f59dd5f0af425ac71<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9809-fe19b55.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9809-fe19b55.dmg

**Comment by PhilipRosedale** [comment]
Love the behavior, works great for me!

**Comment by hifi-gustavo** [comment]
Build console output for commit: 04d3bf0c3802e1a77dee971eca94a915ca6c1dc6 available at: https://gist.github.com/e1058c73d102567f59dd5f0af425ac71<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9809-04d3bf0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9809-04d3bf0.dmg

**Comment by Nex-Pro** [comment]
Testing this.

File: scripts/tutorials/entity_scripts/magneticBlock.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    // Helper for detecting nearby objects near entityProperties, with the scale calculated by the dimensions of the object.
+    function findEntitiesInRange(entityProperties) {
+        var dimensions = entityProperties.dimensions;
+        return Entities.findEntities(entityProperties.position, ((dimensions.x + dimensions.y + dimensions.z) / 3) * RANGE_MULTIPLER);
+    }
+
+    function getNearestValidEntityProperties(releasedProperties) {
+        var entities = findEntitiesInRange(releasedProperties);
+        var nearestEntity = null;
+        var nearest = 9999999999999;
```
Comment (thoys) [inline scripts/tutorials/entity_scripts/magneticBlock.js]: Could use ```Number.MAX_VALUE``` here

File: scripts/tutorials/makeBlocks.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    // Helper functions.
+    function randomPastelRGBComponent() {
+        return Math.floor(Math.random() * MAX_RGB_COMPONENT_VALUE) + MIN_COLOR_VALUE;
+    }
+
+    var SCRIPT_URL = Script.resolvePath("./entity_scripts/magneticBlock.js");
+
+    var frontVector = Quat.getFront(MyAvatar.orientation);
+    frontVector.y -=.25;
```
Comment (thoys) [inline scripts/tutorials/makeBlocks.js]: Magic number

File: scripts/tutorials/makeBlocks.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    for(var x =0; x < 3; x++) {
```
Comment (thoys) [inline scripts/tutorials/makeBlocks.js]: the 3 seems like a magic number, and some styling on this line is preferred

File: scripts/tutorials/makeBlocks.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+            var frontOffset = {
+                x: 0,
+                y: SIZE * y + SIZE,
+                z: SIZE * x + SIZE
+            };
+
+            Entities.addEntity({
+                type: "Box",
+                name: "MagneticBlock-" + y +'-' + x,
```
Comment (thoys) [inline scripts/tutorials/makeBlocks.js]: spacing

File: scripts/tutorials/makeBlocks.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                dimensions: {
+                    x: SIZE,
+                    y: SIZE,
+                    z: SIZE
+                },
+                userData: JSON.stringify({grabbableKey: { cloneable: true, grabbable: true, cloneLifetime : LIFETIME, cloneLimit: 9999}}),
```
Comment (thoys) [inline scripts/tutorials/makeBlocks.js]: spacing

File: scripts/tutorials/entity_scripts/magneticBlock.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    function MagneticBlock() {}
+    // Bind pre-emptive events
+    MagneticBlock.prototype = {
+        // When script is bound to an entity, preload is the first callback called with the entityID. It will behave as the constructor
+        preload: function(id) {
+            /*
+              We will now override any existing userdata with the grabbable property.
+              Only retrieving userData
+            */
+            var val = Entities.getEntityProperties(id, ['userData'])
```
Comment (thoys) [inline scripts/tutorials/entity_scripts/magneticBlock.js]: val? and missing semicolon

**Comment by Nex-Pro** [comment]
### PASS

Works as per the description. 

Feedback:

I scaled one of the cloneable cubes to be ~3x longer in one axis. 
This entity would always snap in the same orientation against another similar entity.   A regular cube would snap centered on any face of the stretched ent. 

Because they are symmetrical the rotation that is taking place is not observed. 

Therefore I could build 3 Entity 'H' shape arrangements with the stretched Ent and other un-stretched Ents  but not with the scaled ones. 

This is because the rotations are being aligned rather than snapped in local rotation of the Ent to snap to + some 90 degree increment. Might be a nice tweak.

File: scripts/tutorials/entity_scripts/magneticBlock.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+(function() {
+    const SNAPSOUND_SOURCE = SoundCache.getSound(Script.resolvePath("../../system/assets/sounds/entitySnap.wav?xrs"));
+    const RANGE_MULTIPLER = 1.5;
+    const MAX_SCALE = 2;
+    const MIN_SCALE = 0.5;
+
+    // Helper for detecting nearby objects near entityProperties, with the scale calculated by the dimensions of the object.
+    function findEntitiesInRange(entityProperties) {
+        var dimensions = entityProperties.dimensions;
+        return Entities.findEntities(entityProperties.position, ((dimensions.x + dimensions.y + dimensions.z) / 3) * RANGE_MULTIPLER);
```
Comment (thoys) [inline scripts/tutorials/entity_scripts/magneticBlock.js]: Maximal 128 characters per line, split it up please. And the 3 seems like a magic number. Please define it

Comment (Menithal) [inline scripts/tutorials/entity_scripts/magneticBlock.js]: Average of all 3 dimensions. Not a magic Val.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 01ba44c572d66be6c86d79390918bda798576ba6 available at: https://gist.github.com/e1058c73d102567f59dd5f0af425ac71<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9809-01ba44c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9809-01ba44c.dmg

**Pull request merged**